adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,355 @@
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    req,
+		Headers: raiHeaders,
+	}
+
+	return []*adapters.RequestData{requestData}, raiErrors
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &bidResp); err != nil {
+
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: "banner",
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func setHeaders(raiHeaders *http.Header) {
+	raiHeaders.Set("Content-Type", "application/json;charset=utf-8")
+	raiHeaders.Set("Accept", "application/json")
+	raiHeaders.Add("X-Openrtb-Version", "2.5")
+}
+
+func setImp(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (resImps []openrtb2.Imp, err error) {
+	//Imp: Id, tagId, Secure, Bidfloor, Bidfloorcur, Banner, test
+	//Banner: Id, W, H, Format
+	//Format: W, H
+
+	for _, imp := range request.Imp {
+		var secure = int8(0)
+		raiExt, errImp := parseImpExt(&imp)
+
+		if errImp != nil {
+			return nil, errImp
+		}
+
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
+
+			if raiExt.TestRa {
+				richaudienceRequest.Test = int8(1)
+			}
+
+			if raiExt.BidFloor <= 0 {
+				imp.BidFloor = 0.00001
+			} else {
+				imp.BidFloor = raiExt.BidFloor
+			}
+
+			if raiExt.BidFloorCur == "" {
+				imp.BidFloorCur = "USD"
+			} else {
+				imp.BidFloorCur = raiExt.BidFloorCur
+			}
+		}
+
+		if request.Site != nil && request.Site.Page != "" {
+			pageURL, error := url.Parse(request.Site.Page)
+			if error == nil && pageURL.Scheme == "https" {
+				secure = int8(1)
+			}
+		}
+		imp.Secure = &secure
+
+		if imp.Banner == nil {
+			err = &errortypes.BadInput{
+				Message: "Banner Object not found",
+			}
+			return nil, err
+		} else {
+			raiBanner := *imp.Banner
```

Comment:
Why do you need to make a copy of the banner object? You are not modifying it. I recommend directly referencing imp.Banner in the checks and remove the `raiBanner` variable.

Reply:
in next commit i have solved